django/db/models/sql/query.py

```diff
@@ -1573,7 +1573,11 @@ def build_filtered_relation_q(
 
     def add_filtered_relation(self, filtered_relation, alias):
         filtered_relation.alias = alias
-        lookups = dict(get_children_from_q(filtered_relation.condition))
+        lookups = dict(
```

Comment:
I think this is only _masking_ the crash by dropping the false criteria

If it was generating a join along the lines of `LEFT JOIN book AS the_book ON (book.author_id = author.id AND false)`, as should be expected in this case, then `the_book.id` should always be `NULL` as `(book.author_id = author.id AND false)` will always be false.

Your test demonstrates that the criteria is dropped and produces wrong results.

The code must be adapted to handle conditional expressions or raise a more meaningful error when provided to `FilteredRelation(condition)` not discard them.

Reply:
Hi @charettes, 

the method I changed drop the false criteria only during the check of lookups. The filtered lookup dict get discarded after the check.

This is the resulting SQL query, according to me it's the expected one:

```SELECT
    "filtered_relation_author"."id",
    "filtered_relation_author"."name",
    "filtered_relation_author"."content_type_id",
    "filtered_relation_author"."object_id"
FROM "filtered_relation_author" INNER JOIN "filtered_relation_book" the_book
ON ("filtered_relation_author"."id" = the_book."author_id" AND (False))
WHERE the_book."id" IS NOT NULL
```

Let me know if I'm missing something 😉

Reply:
I realize I should have made my comment clearer and I misunderstood the test, sorry for that.

When `child` is an instance of `Expression` it could refer to any lookup though (e.g. `Exact('otherfield__outside', 3)`) so the question that remains with this ticket is whether we want to validate that as well.

Reply:
@charettes I got the point.. let me work on it 👍

Reply:
@charettes I have implemented a fix for this case and I have added some tests to increase coverage. By the way, I found a bug when the condition didn't specify an explicit lookup.

Let me know what do you think... thanks!

Reply:
> 

`INNER JOIN "filtered_relation_book" the_book` implicit `the_book."id" IS NOT NULL`